tools/gonilnilfunctions/pkg/analyzer/analyzer.go

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package analyzer
+
+import (
+	"go/ast"
+	"go/types"
+	"strings"
+
+	"golang.org/x/tools/go/analysis"
+	"golang.org/x/tools/go/analysis/passes/inspect"
+	"golang.org/x/tools/go/ast/inspector"
+)
+
+var Analyzer = &analysis.Analyzer{
+	Name:     "gonilnilfunctions",
+	Doc:      "Verifies that every go function with error as one of many return types cannot return nil, nil",
+	Run:      run,
+	Requires: []*analysis.Analyzer{inspect.Analyzer},
+}
+
+// getNestedReturnStatements searches the AST for return statements, and returns
+// them in a tail-call optimized list.
+func getNestedReturnStatements(s ast.Stmt, returns []*ast.ReturnStmt) []*ast.ReturnStmt {
+	switch s := s.(type) {
+	case *ast.BlockStmt:
+		statements := make([]*ast.ReturnStmt, 0)
+		for _, stmt := range s.List {
+			statements = append(statements, getNestedReturnStatements(stmt, make([]*ast.ReturnStmt, 0))...)
+		}
+
+		return append(returns, statements...)
+	case *ast.BranchStmt:
+		return returns
+	case *ast.ForStmt:
+		return getNestedReturnStatements(s.Body, returns)
+	case *ast.IfStmt:
+		return getNestedReturnStatements(s.Body, returns)
+	case *ast.LabeledStmt:
+		return getNestedReturnStatements(s.Stmt, returns)
+	case *ast.RangeStmt:
+		return getNestedReturnStatements(s.Body, returns)
+	case *ast.ReturnStmt:
+		return append(returns, s)
+	case *ast.SwitchStmt:
+		return getNestedReturnStatements(s.Body, returns)
+	case *ast.SelectStmt:
+		return getNestedReturnStatements(s.Body, returns)
+	case *ast.TypeSwitchStmt:
+		return getNestedReturnStatements(s.Body, returns)
+	case *ast.CommClause:
+		statements := make([]*ast.ReturnStmt, 0)
+		for _, stmt := range s.Body {
+			statements = append(statements, getNestedReturnStatements(stmt, make([]*ast.ReturnStmt, 0))...)
+		}
+
+		return append(returns, statements...)
+	case *ast.CaseClause:
+		statements := make([]*ast.ReturnStmt, 0)
+		for _, stmt := range s.Body {
+			statements = append(statements, getNestedReturnStatements(stmt, make([]*ast.ReturnStmt, 0))...)
+		}
+
+		return append(returns, statements...)
+	case *ast.ExprStmt:
+		return returns
+	}
+	return returns
+}
+
+// run runs the analysis. Ironically, it also fails its own check,
+// so we'll skip analysis using the following invocation:
+// ignore-nil-nil-function-check
```

Comment:
If we wanted to be post-ironic, we could define a `success` type (maybe an empty struct?), use `ResultType: reflect.TypeOf(success{})` and modify this method to return `success{}`.

I'm not sure if it's worth it, or if it would break anything (I assume not), but just a thought that occurred to me while looking at this.

Reply:
Perhaps `success` isn't quite right for the name, but you get the idea.